src/ResourceManager/Kubernetes/Commands.Kubernetes/Commands/CreateOrUpdateKubeBase.cs

```diff
@@ -0,0 +1,363 @@
+using System.IO;
+using System.Linq;
+using System.Management.Automation;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.RegularExpressions;
+using System.Threading;
+using Microsoft.Azure.Commands.Kubernetes.Generated;
+using Microsoft.Azure.Commands.Kubernetes.Generated.Models;
+using Microsoft.Azure.Commands.Kubernetes.Models;
+using Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters;
+using Microsoft.Azure.Commands.ResourceManager.Common.Tags;
+using Microsoft.Azure.Graph.RBAC.Version1_6;
+using Microsoft.Azure.Graph.RBAC.Version1_6.Models;
+using Microsoft.Azure.Management.Authorization.Version2015_07_01;
+using Microsoft.Azure.Management.Authorization.Version2015_07_01.Models;
+using Microsoft.Azure.Management.Internal.Resources;
+using Microsoft.Rest.Azure;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Commands.Kubernetes
+{
+    [Cmdlet(VerbsCommon.New, KubeNounStr, DefaultParameterSetName = DefaultParamSet)]
+    [OutputType(typeof(PSObject), typeof(List<PSObject>))]
+    public abstract class CreateOrUpdateKubeBase : KubeCmdletBase
+    {
+        protected const string DefaultParamSet = "defaultParameterSet";
+        protected const string SpParamSet = "servicePrincipalParameterSet";
+        protected readonly Regex _dnsRegex = new Regex("[^A-Za-z0-9-]");
+
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ParameterSetName = DefaultParamSet,
+            HelpMessage = "Resource Group Name.")]
+        [Parameter(
+            Position = 0,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "Resource Group Name.")]
+        [ResourceGroupCompleter()]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ParameterSetName = DefaultParamSet,
+            HelpMessage = "Kubernetes managed cluster Name.")]
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "Kubernetes managed cluster Name.")]
+        [ValidateNotNullOrEmpty]
+        [ValidatePattern("^[a-zA-Z0-9][a-zA-Z0-9_.-]*$")]
+        [ValidateLength(2, 64)]
+        public string Name { get; set; }
+
+        [Parameter(
+            Position = 2,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage =
+                "The client ID of the AAD application / service principal used for cluster authentication to Azure APIs.")]
+        public string ClientId { get; set; }
+
+        [Parameter(
+            Position = 3,
+            Mandatory = true,
+            ParameterSetName = SpParamSet,
+            HelpMessage = "The secret associated with the AAD application / service principal.")]
+        public string ClientSecret { get; set; }
```

Comment:
This should be a secure string, or ClientId + ClientSecret could be represented as a PSCredetial